sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala

```diff
@@ -0,0 +1,190 @@
+
+  private def check(normalizedFormat: String, numberFormat: String) = {
+    def invalidSignPosition(format: String, c: Char): Boolean = {
+      val signIndex = format.indexOf(c)
+      signIndex > 0 && signIndex < format.length - 1
+    }
+
+    if (normalizedFormat.count(_ == pointSign) > 1) {
+      fail(s"'$letterPointSign' or '$pointSign'", numberFormat)
+    } else if (normalizedFormat.count(_ == minusSign) > 1) {
+      fail(s"'$letterMinusSign' or '$minusSign'", numberFormat)
+    } else if (normalizedFormat.count(_ == dollarSign) > 1) {
+      fail(s"'$dollarSign'", numberFormat)
+    } else if (invalidSignPosition(normalizedFormat, minusSign)) {
+      throw new AnalysisException(
+        s"'$letterMinusSign' or '$minusSign' must be the first or last char")
+    } else if (invalidSignPosition(normalizedFormat, dollarSign)) {
+      throw new AnalysisException(s"'$dollarSign' must be the first or last char")
+    }
+  }
+
+  /**
+   * Convert string to numeric based on the given number format.
+   * The format can consist of the following characters:
+   * '9':  digit position (can be dropped if insignificant)
+   * '0':  digit position (will not be dropped, even if insignificant)
+   * '.':  decimal point (only allowed once)
+   * ',':  group (thousands) separator
+   * 'S':  sign anchored to number (uses locale)
+   * 'D':  decimal point (uses locale)
+   * 'G':  group separator (uses locale)
+   * '$':  specifies that the input value has a leading $ (Dollar) sign.
+   *
+   * @param input the string need to converted
+   * @param numberFormat the given number format
+   * @return decimal obtained from string parsing
+   */
+  def parse(input: UTF8String, numberFormat: String): Decimal = {
+    val normalizedFormat = normalize(numberFormat)
+    check(normalizedFormat, numberFormat)
+
+    val precision = normalizedFormat.filterNot(isSign).length
+    val formatSplits = normalizedFormat.split(pointSign)
+    val scale = if (formatSplits.length == 1) {
+      0
+    } else {
+      formatSplits(1).filterNot(isSign).length
+    }
+    val transformedFormat = transform(normalizedFormat)
+    val numberFormatInstance = NumberFormat.getInstance()
+    val numberDecimalFormat = numberFormatInstance.asInstanceOf[DecimalFormat]
+    numberDecimalFormat.setParseBigDecimal(true)
+    numberDecimalFormat.applyPattern(transformedFormat)
+    val parsePosition = new ParsePosition(0)
+    val inputStr = input.toString.trim
+    val inputSplits = inputStr.split(pointSign)
+    if (inputSplits.length == 1) {
+      if (inputStr.filterNot(isSign).length > precision - scale) {
+        throw QueryExecutionErrors.invalidNumberFormatError(numberFormat)
+      }
+    } else if (inputSplits(0).filterNot(isSign).length > precision - scale ||
+      inputSplits(1).filterNot(isSign).length > scale) {
+      throw QueryExecutionErrors.invalidNumberFormatError(numberFormat)
+    }
+    val number = numberDecimalFormat.parse(inputStr, parsePosition)
```

Comment:
`parsePosition` can be inlined since it is used only once:
```suggestion
    val number = numberDecimalFormat.parse(inputStr, new ParsePosition(0))
```